### PR TITLE
SAAS-2316: Instance Template Machine Types for Google Compute Engine

### PR DESCRIPTION
**Pull request by mehakseedat63**

### CODE REVIEW

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+var async = require('async');
+var helpers = require('../../../helpers/google');
+
+module.exports = {
+    title: 'Instance Template Machine Type',
+    category: 'Compute',
+    description: 'Ensures that Google Cloud Virtual Machine instances are of given types.',
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:7]:
```suggestion
    description: 'Ensure that Cloud VM instance templates are of given types.',
```

Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:7]:
```suggestion
    description: 'Ensure that Cloud Virtual Machine instance templates are of given types.',
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    more_info: 'Google Cloud Virtual Machine instance should be of the given types to ensure the internal compliance and prevent unexpected billing charges.',
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:8]:
```suggestion
    more_info: 'Virtual Machine instance templates should be of the given types to ensure the internal compliance and prevent unexpected billing charges.',
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    link: 'https://cloud.google.com/compute/docs/machine-types',
+    recommended_action: 'Stop the Google Cloud Virtual Machine instance, change the machine type to the desired type  and restart the instance.',
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:10]: Change this as well

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    },
+    run: function (cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions();
+
+        let projects = helpers.addSource(cache, source,
+            ['projects', 'get', 'global']);
+        var instance_template_machine_types = settings.instance_template_machine_types || this.settings.instance_template_machine_types.default;
+        if (!instance_template_machine_types.length) return callback();
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:28]: Move this to line 25

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                    'Unable to query instance templates', region, null, null, instanceTemplates.err);
+                return rcb();
+            }
+
+            if (!instanceTemplates.data.length) {
+                helpers.addResult(results, 0, 'No instance templates found', region);
+                return rcb();
+            }
+            instanceTemplates.data.forEach(instanceTemplate => {
+                if (instanceTemplate.properties.machineType && instance_template_machine_types.includes(instanceTemplate.properties.machineType)) {
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:52]:
```suggestion
                if (instanceTemplate.properties && instanceTemplate.properties.machineType && 
                    instance_template_machine_types.includes(instanceTemplate.properties.machineType)) {
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                    helpers.addResult(results, 0,
+                        'Google Cloud Virtual Machine instance template has the desired machine type', region);
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:54]:
```suggestion
                        'Virtual Machine instance template has desired machine type', region);
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+                } else {
+                    helpers.addResult(results, 2,
+                        'Google Cloud Virtual Machine instance template does not have the desired machine type', region);
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:57]:
```suggestion
                        'Virtual Machine instance template does not have desired machine type', region);
```

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                                    "kind": "compute#attachedDisk"
+                                }
+                            ],
+                            "metadata": {
+                                "fingerprint": "8czNZABQPuc=",
+                                "kind": "compute#metadata"
+                            },
+                            "serviceAccounts": [
+                                {
+                                    "email": "779980017373-compute@developer.gserviceaccount.com",
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:209]: We never want to make account details public
```suggestion
                                    "email": "111122223333@developer.gserviceaccount.com",
```

Comment (AkhtarAmir) [reply plugins/google/compute/InstanceTemplateMachineTypes.spec.js:209]: You can also remove unnecessary data as well. Maybe just keep the data which is needed for plugin

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:116]:
```suggestion
                                    "email": "111122223333@developer.gserviceaccount.com",
```

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+                            "shieldedInstanceConfig": {
+                                "enableSecureBoot": false,
+                                "enableVtpm": false,
+                                "enableIntegrityMonitoring": false
+                            },
+                            "confidentialInstanceConfig": {
+                                "enableConfidentialCompute": false
+                            }
+                        },
+                        "selfLink": "https://www.googleapis.com/compute/v1/projects/akhtar-dev-aqua/global/instanceTemplates/instance-template-1",
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:144]:
```suggestion
```

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+                        "kind": "compute#instanceTemplate"
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:238]:
```suggestion
                        },
                        "kind": "compute#instanceTemplate"
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
+var async = require('async');
+var helpers = require('../../../helpers/google');
+
+module.exports = {
+    title: 'Instance Template Machine Type',
+    category: 'Compute',
+    description: 'Ensure that Cloud Virtual Machine instance templates are of given types.',
+    more_info: 'Virtual Machine instance templates should be of the given types to ensure the internal compliance and prevent unexpected billing charges.',
+    link: 'https://cloud.google.com/compute/docs/machine-types',
+    recommended_action: 'Stop the Cloud Virtual Machine instance template, change the machine type to the desired type  and restart the instance template.',
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:10]:
```suggestion
    recommended_action: 'Stop the Cloud Virtual Machine instance template, change the machine type to the desired type and restart the instance template.',
```

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+            const cache = createCache(
+                [],
+                null
+            );
+
+            plugin.run(cache, settings, callback);
+        });
+
+        it('should FAIL if VM instance is not of the desired machine type', function (done) {
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:61]:
```suggestion
        it('should FAIL if VM instance template does not have desired machine type', function (done) {
```

File: plugins/google/compute/InstanceTemplateMachineTypes.spec.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                        },
+                    }
+                ],
+                null
+            );
+
+            plugin.run(cache, settings, callback);
+        })
+
+        it('should PASS if VM instance is of the desired machine type', function (done) {
```
Comment (AkhtarAmir) [inline plugins/google/compute/InstanceTemplateMachineTypes.spec.js:88]:
```suggestion
        it('should PASS if VM instance template has desired machine type', function (done) {
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
+var async = require('async');
+var helpers = require('../../../helpers/google');
+
+module.exports = {
+    title: 'Instance Template Machine Type',
+    category: 'Compute',
+    description: 'Ensure that Cloud Virtual Machine instance templates are of given types.',
+    more_info: 'Virtual Machine instance templates should be of the given types to ensure the internal compliance and prevent unexpected billing charges.',
+    link: 'https://cloud.google.com/compute/docs/machine-types',
+    recommended_action: 'Stop the Cloud Virtual Machine instance template, change the machine type to the desired type and restart the instance template.',
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:10]:
```suggestion
    recommended_action: 'Ensure that Virtual Machine instance template are not using undesired machine types.',
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+            description: 'Desired Google Cloud Virtual Machine instance template type',
+            regex: '^.*$',
+            default: ''
+        },
+    },
+    run: function (cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions();
+        var instance_template_machine_types = settings.instance_template_machine_types || this.settings.instance_template_machine_types.default;
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:24]:
```suggestion
        var regions = helpers.regions();

        var instance_template_machine_types = settings.instance_template_machine_types || this.settings.instance_template_machine_types.default;
```

File: plugins/google/compute/instanceTemplateMachineTypes.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+        if (!instance_template_machine_types.length) return callback();
+        let projects = helpers.addSource(cache, source,
+            ['projects', 'get', 'global']);
```
Comment (AkhtarAmir) [inline plugins/google/compute/instanceTemplateMachineTypes.js:28]:
```suggestion
        var instance_template_machine_types = settings.instance_template_machine_types || this.settings.instance_template_machine_types.default;

        if (!instance_template_machine_types.length) return callback(null, results, source);

        let projects = helpers.addSource(cache, source,
            ['projects', 'get', 'global']);
```

**Review by giorod3 — APPROVED** [review]

**Pull request merged**